spec/requests/rack_attack_spec.rb

```diff
@@ -0,0 +1,72 @@
+require "spec_helper"
+
+describe "Rack::Attack" do
+  before { freeze_time }
+
+  def unique_ip_env
+    { "REMOTE_ADDR" => Faker::Internet.unique.ip_v4_address }
+  end
+
+  def unique_user_params
+    { user: { login: Faker::Name.unique.first_name, password: "secret" } }
+  end
+
+  it "test utility returns valid parameters for successful user login attempts" do
+    params = unique_user_params
+    create(:user, login: params[:user][:login], password: params[:user][:password])
+    post user_session_path, params: params.to_query
+    expect(response).to have_http_status(:redirect)
+  end
+
+  context "when there have been max user login attempts from an IP address" do
+    let(:ip) { Faker::Internet.unique.ip_v4_address }
+
+    before do
+      ArchiveConfig.RATE_LIMIT_LOGIN_ATTEMPTS.times do
+        post user_session_path, params: unique_user_params.to_query, env: { "REMOTE_ADDR" => ip }
+      end
+    end
+
+    it "throttles the next attempt from the same IP" do
+      post user_session_path, params: unique_user_params.to_query, env: { "REMOTE_ADDR" => ip }
+      expect(response).to have_http_status(:too_many_requests)
+    end
+
+    it "does not throttle the next attempt from the same IP after some time" do
+      travel ArchiveConfig.RATE_LIMIT_LOGIN_PERIOD.seconds
+      post user_session_path, params: unique_user_params.to_query, env: { "REMOTE_ADDR" => ip }
+      expect(response).to have_http_status(:ok)
+    end
+
+    it "does not throttle an attempt from a different IP" do
```

Comment:
Could I be annoying and suggest flipping the order, so `it "does not throttle an attempt from a different IP"` is above `it "does not throttle the next attempt from the same IP after some time"`? I think that way it might be clearer that the attempt from a different IP is coming before the rate limiting period is up.